pkg/controller/persistentvolume/persistentvolume_provisioner_controller.go

```diff
@@ -0,0 +1,489 @@
+	newVolume.Spec.ClaimRef = claimRef
+	newVolume.Annotations[pvProvisioningRequiredAnnotationKey] = "true"
+	newVolume.Annotations[qosProvisioningKey] = storageClass
+	newVolume, err = controller.client.CreatePersistentVolume(newVolume)
+	glog.V(5).Infof("Unprovisioned PersistentVolume[%s] created for PVC[%s], which will be fulfilled in the storage provider", newVolume.Name, claim.Name)
+	if err != nil {
+		return fmt.Errorf("PersistentVolumeClaim[%s] failed provisioning: %+v", claim.Name, err)
+	}
+
+	claim.Annotations[pvProvisioningRequiredAnnotationKey] = pvProvisioningCompletedAnnotationValue
+	_, err = controller.client.UpdatePersistentVolumeClaim(claim)
+	if err != nil {
+		glog.Error("error updating persistent volume claim: %v", err)
+	}
+
+	return nil
+}
+
+func (controller *PersistentVolumeProvisionerController) reconcileVolume(pv *api.PersistentVolume) error {
+	glog.V(5).Infof("PersistentVolume[%s] reconciling", pv.Name)
+
+	if pv.Spec.ClaimRef == nil {
+		glog.V(5).Infof("PersistentVolume[%s] is not bound to a claim.  No provisioning required", pv.Name)
+		return nil
+	}
+
+	// TODO:  fix this leaky abstraction.  Had to make our own store key because ClaimRef fails the default keyfunc (no Meta on object).
+	obj, exists, _ := controller.claimStore.GetByKey(fmt.Sprintf("%s/%s", pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name))
+	if !exists {
+		return fmt.Errorf("PersistentVolumeClaim[%s/%s] not found in local cache", pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name)
+	}
+
+	claim := obj.(*api.PersistentVolumeClaim)
+
+	// no provisioning required, volume is ready and Bound
+	if !keyExists(pvProvisioningRequiredAnnotationKey, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolume[%s] does not require provisioning", pv.Name)
+		return nil
+	}
+
+	// provisioning is completed, volume is ready.
+	if isAnnotationMatch(pvProvisioningRequiredAnnotationKey, pvProvisioningCompletedAnnotationValue, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolume[%s] is bound and provisioning is complete", pv.Name)
+		if pv.Spec.ClaimRef.Namespace != claim.Namespace || pv.Spec.ClaimRef.Name != claim.Name {
+			return fmt.Errorf("pre-bind mismatch - expected %s but found %s/%s", claimToClaimKey(claim), pv.Spec.ClaimRef.Namespace, pv.Spec.ClaimRef.Name)
+		}
+		return nil
+	}
+
+	// provisioning is incomplete.  Attempt to provision the volume.
+	if !isAnnotationMatch(pvProvisioningRequiredAnnotationKey, pvProvisioningCompletedAnnotationValue, pv.Annotations) {
+		glog.V(5).Infof("PersistentVolume[%s] provisioning in progress", pv.Name)
+		provisionVolume(pv, controller)
+	}
+
+	return nil
+}
+
+// provisionVolume provisions a volume that has been created in the cluster but not yet fulfilled by
+// the storage provider.
+func provisionVolume(pv *api.PersistentVolume, controller *PersistentVolumeProvisionerController) {
+	if isAnnotationMatch(pvProvisioningRequiredAnnotationKey, pvProvisioningCompletedAnnotationValue, pv.Annotations) {
```

Comment:
I think we repeat this in enough places that it should have a utility method `isProvisioningComplete(pv)`